core/src/main/java/software/amazon/awssdk/SdkRequestOverrideConfig.java

```diff
@@ -0,0 +1,131 @@
+ */
+public abstract class SdkRequestOverrideConfig<B extends SdkRequestOverrideConfig.Builder<B, C>,
+        C extends SdkRequestOverrideConfig<B, C>> implements ToCopyableBuilder<B, SdkRequestOverrideConfig<B, C>> {
+
+    private final ProgressListener progressListener;
+
+    private final Map<String, String> customHeaders;
+
+    private final Map<String, List<String>> customQueryParameters;
+
+    protected SdkRequestOverrideConfig(B builder) {
+        this.progressListener = builder.getProgressListener();
+        this.customHeaders = builder.getCustomHeaders();
+        this.customQueryParameters = builder.getCustomQueryParameters();
+    }
+
+    public Optional<ProgressListener> progressListener() {
+        return Optional.ofNullable(progressListener);
+    }
+
+    public Optional<Map<String, String>> customHeaders() {
+        return Optional.ofNullable(customHeaders);
+    }
+
+    public Optional<Map<String, List<String>>> customQueryParameters() {
+        return Optional.ofNullable(customQueryParameters);
+    }
+
+    public interface Builder<B extends SdkRequestOverrideConfig.Builder<B, C>,
+            C extends SdkRequestOverrideConfig<B, C>> extends CopyableBuilder<B, SdkRequestOverrideConfig<B, C>> {
+        ProgressListener getProgressListener();
+
+        B progressListener(ProgressListener progressListener);
+
+        Map<String, String> getCustomHeaders();
+
+        B customHeaders(Map<String, String> customHeaders);
+
+        Map<String, List<String>> getCustomQueryParameters();
+
+        B customQueryParameters(Map<String, List<String>> customQueryParameters);
+    }
+
+    protected static abstract class BuilderImpl<B extends SdkRequestOverrideConfig.Builder<B, C>,
+            C extends SdkRequestOverrideConfig<B, C>> implements Builder<B, C> {
+        private final Class<B> concrete;
+
+        private ProgressListener progressListener;
+
+        private Map<String, String> customHeaders;
+
+        private Map<String, List<String>> customQueryParameters;
+
+        protected BuilderImpl(Class<B> concrete) {
+            Validate.isInstanceOf(concrete, this, "This class is not an instance of %s!", concrete);
+            this.concrete = concrete;
+        }
+
+        protected BuilderImpl(Class<B> concrete, C sdkRequestOverrideConfig) {
+            this(concrete);
+            sdkRequestOverrideConfig.progressListener().map(this::progressListener);
+            sdkRequestOverrideConfig.customHeaders().map(this::customHeaders);
+            sdkRequestOverrideConfig.customQueryParameters().map(this::customQueryParameters);
+        }
+
+        @Override
+        public ProgressListener getProgressListener() {
+            return progressListener;
+        }
+
+        @Override
+        public B progressListener(ProgressListener progressListener) {
+            this.progressListener = progressListener;
+            return concrete.cast(this);
```

Comment:
We usually just do an unchecked cast with the other builders. Should we do this approach there, or that approach here?

Reply:
Don't really have a preference. The only reason I chose to go with using this method was to avoid need to suppress the unchecked warning

Reply:
I don't really have a preference either.